first commit

iliabv opened the pull request.

Commit:
first commit

acvetkov commented on test/.eslintrc:
```diff
@@ -0,0 +1,9 @@
+     "sinon": false,
+     "assert": false,
+   }
+}
```
new line required :)

acvetkov commented:
- replace `var` by `let` or `const`
- remove unnecessary dependencies
- rename module in `package.json` (logger module name is already taken)
- append modules overview and jsdoc

vitalets commented:
Please, describe how it works..

vitalets commented on src/console/index.js:
```diff
@@ -0,0 +1,60 @@
+    constructor (replacer = createReplacer()) {
+        this.replacePlaceholders = false;
+        this._replacer = replacer;
+        this._console = console ? console : createFakeConsole();
```
what is fake console?

iliabv replied:
It's for environments without global console object

iliabv replied:
I've just realized, that It can be replaces with simple `{}`

vitalets commented on src/console/index.js:
```diff
@@ -0,0 +1,60 @@
+
+    write (method, formattedArgs) {
+        if (!this._console[method]) {
+            return;
```
throw error?

iliabv replied:
It's possible, that just one transport would support particular method, and others would ignore it. On the other hand, some transports doesn't even look at `method` value, like `WebFileTransport`

vitalets commented on src/console/index.js:
```diff
@@ -0,0 +1,60 @@
+        if (prefix) {
+            args = appendPrefix(prefix, args);
+        }
+        if (this.replacePlaceholders) {
```
I prefer

``` js
return this.replacePlaceholders 
    ? replacePlaceholders(this._replacer, args)
    : args
```

vitalets commented on src/index.js:
```diff
@@ -0,0 +1,25 @@
+logger.ConsoleTransport = ConsoleTransport;
+logger.WebFileTransport = WebFileTransport;
+
+export default logger;
```
why we create logger instance here?

vitalets commented on src/console/index.js:
```diff
@@ -0,0 +1,60 @@
+
+import Replacer from '../replacer';
+
+export default class ConsoleTransport {
```
I like when file is named like underlying class..

vitalets commented on src/logger.js:
```diff
@@ -0,0 +1,60 @@
+            });
+    }
+
+    setLevels (transport, levels) {
```
isn't it better to move level configuration into each transport? And logger is just a hub..

iliabv replied:
I wanted to make transports as simple as possible. Just objects with `format` and `write` methods. It's simpler to create new transports this way, without thinking about levels and stuff

Commit:
Add "How It Works" to readme

Commit:
Update readme

ilya-babanov commented:
https://github.com/ilya-babanov/logger/blob/second/README.md

Commit:
Remove fakeConsole from console transport, improve webFile transport

Commit:
Add new line to .eslintrc

Commit:
Rename replacer class

vitalets commented on src/logger.js:
```diff
@@ -0,0 +1,60 @@
+}
+
+function isMethodAllowed(method, levels) {
+    if (Array.isArray(levels)) {
```
``` js
return Array.isArray(levels)
    ? levels.indexOf(method) !== -1
    : true;
```

vitalets commented on src/replacer.js:
```diff
@@ -0,0 +1,78 @@
+
+function formatIntegerNumber(nanString, data) {
+    return typeof data === 'number' ?
+        String(Math.floor(data)) : nanString;
```
maybe Math.round() ?

vitalets commented on src/replacer.js:
```diff
@@ -0,0 +1,78 @@
+}
+
+function formatJSON(data) {
+    return JSON.stringify(data || '', null, 1);
```
try catch

vitalets commented on src/web-file/file.js:
```diff
@@ -0,0 +1,92 @@
+            return;
+        }
+        try {
+            var requestFs = window.webkitRequestFileSystem || window.requestFileSystem;
```
maybe change order?
`var requestFs = window.requestFileSystem || window.webkitRequestFileSystem;`

iliabv replied:
There is no `requestFileSystem` in Chrome, only `webkitRequestFileSystem`

vitalets replied:
..so in chrome it will take `webkitRequestFileSystem`.
Unprefixed globals are preferred over prefixed

vitalets commented on test/environment.js:
```diff
@@ -0,0 +1,39 @@
+
+import chai from 'chai';
+import chaiAsPromised from 'chai-as-promised';
+import sinonChrome from 'sinon-chrome';
```
remove

Commit:
Add jsdocs

Commit:
Remove concept of transports

Add writers and formatters

Commit:
Update readme

Commit:
web-file refactoring

Commit:
Improve logger jsdocs

Commit:
Add assertions to Logger constructor

Commit:
Add logic for synchronising parent->clone options

acvetkov commented on src/logger.js:
```diff
@@ -0,0 +1,171 @@
+        if (!options[optionName]) {
+            throw new Error(
+                'You must specify "options.' + optionName + '" parameter for Logger'
+            );
```
Template string instead of string concatination

Commit:
Add tests for replacer

Commit:
remove sinon-chrome dependency

Commit:
Add 'fork' method to logger

Commit:
Fix index.js and README

Commit:
Add examples of cloned and forked loggers to readme

Commit:
Add tests for utils module

Commit:
Fix utils.shallowCopyObject test

Commit:
Add tests for console/writer

Commit:
Fix console/formatter logic, fix writer tests, add formatter tests

Commit:
Improve web-file file.js and formatter.js

Commit:
Fix web-file/file, change web-file formatter

Commit:
Add console normalizer module

Commit:
Improve replacer

vitalets commented on README.md:
````diff
@@ -0,0 +1,43 @@
+# How It Works
+
+```javascript
+import {Logger, ConsoleWriter, ConsoleFormatter, WebFileFriter, WebFileFormatter} from 'logger';
````
typo `WebFileFriter`

vitalets commented on src/console/writer.js:
```diff
@@ -0,0 +1,20 @@
+     * @param {Array<*>} formattedArgs
+     */
+    write (method, prefix, formattedArgs) {
+        if (this._console[method]) {
```
will not work in IE. We should explicitly call console.error(), console.info(), etc

iliabv replied:
Fixed

vitalets commented on src/index.js:
```diff
@@ -0,0 +1,23 @@
+ * Entry point of logger package
+ */
+
+import Logger from './logger';
```
``` js
export {default as Logger} from './logger';
```

iliabv replied:
Fixed

vitalets commented on src/logger.js:
```diff
@@ -0,0 +1,179 @@
+     * @param {} method
+     * @param {} args
+     */
+    _updateClones (method, args) {
```
more es6 way

``` js
_updateClones (method, ...args) {
```

vitalets commented on src/logger.js:
```diff
@@ -0,0 +1,179 @@
+ * @returns {Boolean}
+ */
+function isMethodAllowed(method, levels) {
+    if (Array.isArray(levels)) {
```
``` js
return Array.isArray(levels) ? levels.indexOf(method) !== -1 : true;
```

iliabv replied:
But I like current style more

vitalets commented on src/logger.js:
```diff
@@ -0,0 +1,179 @@
+        throw new Error('You must specify "options" parameter for Logger');
+    }
+    REQUIRED_OPTIONS.forEach(name => {
+        if (!options[name]) {
```
to support boolean options

``` js
if (options[name] === undefined) {
```

vitalets commented on src/replacer.js:
```diff
@@ -0,0 +1,133 @@
+    try {
+        return JSON.stringify(data || '', null, 1);
+    } catch (e) {
+        return '';
```
silent suppress of error is not good. Better to write it into log

``` js
return String(e);
```

vitalets commented on src/utils.js:
```diff
@@ -0,0 +1,48 @@
+    if (!prefix) {
+        return args;
+    }
+    if (typeof args[0] === 'string') {
```
pls describe.. why we should check if args[0] is string

iliabv replied:
It's for consistency with `replacePlaceholders`, `format` and `write` methods.
`formatter.format` can return whatever `writer.write` can handle, but all included writers and formatters work with arrays. This arrays can be of two 'types': array of raw log arguments or array of one formatted argument as a string.

vitalets replied:
it definitly should be described somewhere in source code..

vitalets commented on test/environment.js:
```diff
@@ -0,0 +1,39 @@
+import chaiAsPromised from 'chai-as-promised';
+import sinon from 'sinon';
+
+export function set() {
```
why we export this function? is it used somewhere?

Commit:
Update readme

Commit:
Improve logger._updateClones method

Commit:
Write json parse errors to %j/%o placeholders

Commit:
Update README

Commit:
Attempt to use requestFileSystem before prefixed version

Commit:
Simplify test/environment

Commit:
Use apply to call console methods

Commit:
Change project name to lggr

Commit:
Add jsdocs to web-file/file

acvetkov commented:
I like it :+1:

Commit:
Improve logger.fork method

Use this.constructor parameter for creating new logger.
It's simplifyes extension of Logger class

Commit:
Add tests for new logic of logger.fork

ilya-babanov commented:
I think it would be OK to merge this PR right now. 
Additional notes can always be delivered to me directly and I will try to fix them in future PRs.
